backend/src/apiserver/storage/pipeline_store_kubernetes.go

```diff
@@ -0,0 +1,633 @@
+	pipelineGVR := v2beta1.GroupVersion.WithResource("pipelines")
+
+	var pipelineClient dynamic.ResourceInterface
+	if namespace == "" {
+		pipelineClient = k.noCacheClient.Resource(pipelineGVR)
+	} else {
+		pipelineClient = k.noCacheClient.Resource(pipelineGVR).Namespace(namespace)
+	}
+
+	unstructPipelines, err := pipelineClient.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, util.NewInternalServerError(err, "Failed to list the pipelines")
+	}
+
+	for _, unstructPipeline := range unstructPipelines.Items {
+		if string(unstructPipeline.GetUID()) == pipelineId {
+			k8sPipeline := &v2beta1.Pipeline{}
+			err := runtime.DefaultUnstructuredConverter.FromUnstructuredWithValidation(
+				unstructPipeline.Object, k8sPipeline, true,
+			)
+			if err != nil {
+				return nil, util.NewInternalServerError(err, "Received an invalid pipeline")
+			}
+
+			return k8sPipeline, nil
+		}
+	}
+
+	return nil, util.NewResourceNotFoundError("Pipeline", pipelineId)
+}
+
+func (k *PipelineStoreKubernetes) getK8sPipelineVersions(
+	ctx context.Context, pipelineId string, pipelineVersionId string,
+) (*v2beta1.PipelineVersionList, error) {
+	pipelineVersions := v2beta1.PipelineVersionList{}
+
+	// Be careful, the deep copy is disabled here to reduce memory allocations
+	listOptions := []ctrlclient.ListOption{ctrlclient.UnsafeDisableDeepCopy}
+
+	var namespace string
+
+	if !common.IsMultiUserMode() && common.GetPodNamespace() != "" {
+		namespace = common.GetPodNamespace()
+		listOptions = append(listOptions, ctrlclient.InNamespace(namespace))
+	}
+
+	var errMsg string
+	if pipelineVersionId != "" {
+		errMsg = "Failed to get the pipeline version with ID " + pipelineVersionId
+	} else {
+		errMsg = "Failed to list pipeline versions"
+	}
+
+	if pipelineId != "" {
+		errMsg += " associated with the pipeline with ID " + pipelineId
+		listOptions = append(listOptions, ctrlclient.MatchingLabels{"pipelines.kubeflow.org/pipeline-id": pipelineId})
+	}
+
+	err := k.client.List(ctx, &pipelineVersions, listOptions...)
+	if err != nil {
+		return nil, util.NewInternalServerError(err, errMsg)
+	}
+
+	// If there is no pipeline version ID filter, then just return the results
+	if pipelineVersionId == "" {
+		return &pipelineVersions, nil
+	}
+
+	for _, pipelineVersion := range pipelineVersions.Items {
+		if string(pipelineVersion.UID) == pipelineVersionId {
+			return &v2beta1.PipelineVersionList{Items: []v2beta1.PipelineVersion{pipelineVersion}}, nil
+		}
+	}
+
+	// Fallback to not using the cache if the specific pipeline version is missing
+	pvGVR := v2beta1.GroupVersion.WithResource("pipelineversions")
+
+	var pvClient dynamic.ResourceInterface
+	if namespace == "" {
+		pvClient = k.noCacheClient.Resource(pvGVR)
+	} else {
+		pvClient = k.noCacheClient.Resource(pvGVR).Namespace(namespace)
+	}
+
+	unstructListOptions := metav1.ListOptions{}
+	if pipelineId != "" {
+		// TODO: Verify this is the correct format for a string label selector
```

Comment:
@rimolive did you verify this to do that I had left in the diff?

Reply:
That's a good point. Looks like it's good, given that I did not see any errors during pipeline upload (I tested in my testing cluster) and in the integration tests I did not see any flakiness, so looks like this is working.

Reply:
@rimolive it only gets called when the cache returned not found. So a small Go file that just queries pipeline versions with this code snippet should suffice to confirm it works.